Simplified/Reader2/AdobeDRM/AdobeDRMContentProtection.swift

```diff
@@ -0,0 +1,45 @@
+//
+//  AdobeDRMContentProtection.swift
+//  Simplified
+//
+//  Created by Vladimir Fedorov on 20.01.2021.
+//  Copyright © 2021 NYPL Labs. All rights reserved.
+//
+
+import Foundation
+import R2Shared
+
+#if FEATURE_DRM_CONNECTOR
+
+class AdobeDRMContentProtection: ContentProtection {
+  func open(asset: PublicationAsset, fetcher: Fetcher, credentials: String?, allowUserInteraction: Bool, sender: Any?, completion: @escaping (CancellableResult<ProtectedAsset?, Publication.OpeningError>) -> Void) {
+    
+    // Publication asset must be a FileAsset, as we are opening a file
+    // If not, we can't open and use Adobe DRM
+    guard let fileAsset = asset as? FileAsset else {
+      completion(.failure(.unavailable(nil)))
+      return
+    }
+    
+    do {
+      // META-INF is a part of .epub structure
+      let encryptionPath = "META-INF/encryption.xml"
+      let resource = fetcher.get(encryptionPath)
+      let encryptionData = try resource.read().get()
+      let adobeFetcher = AdobeDRMFetcher(url: fileAsset.url, fetcher: fetcher, encryptionData: encryptionData)
+      let protectedAsset = ProtectedAsset(
+        asset: asset,
+        fetcher: adobeFetcher,
+        onCreatePublication: nil
+      )
+      completion(.success(protectedAsset))
+    } catch {
+      // .success(nil) means it is not an asset protected with Adobe DRM
+      // Streamer continues to iterate over available ContentProtections
+      completion(.success(nil))
```

Comment:
so this means that no matter the reason why `resource.read().get()` throws, here we are saying "there's no valid Adobe DRM, but there IS an encrypted asset that someone else should decrypt" ?  And basically "Failure" means "there's no encrypted asset" ?

Reply:
not exactly. this function is called once, when we open a book. at this moment Streamer doesn't know what Content Protection can work with it, and iterates over Content Protections it has (the app initialise Streamer passing an array of Content Protection classes). If the code in do..catch fails, it means it couldn't initialise AdobeDRMFetcher, but there may be other types of Content Protection that work with it. If we return `.failure` here, Streamer stops iterating and shows the error we pass. It means if the book is not protected, Streamer will show the error and won't try to open it without protection. When Streamer receives `.success` with `nil` value for `ProtectedAsset`, it goes to the next Content Protection.

In short, `.failure` means "this is an adobe DRM protected item, but we can't decrypt it for some reason". `catch` block is called when we have no DRM at all.

Reply:
Would `resource.read().get()` comes back with valid data if the resource is LCP ePub?
If that's the case, we will not be calling `completion(nil)`

Reply:
> Would resource.read().get() comes back with valid data if the resource is LCP ePub?

No, LCP `.epub`s don't contain `META-INF/encryption.xml`. LCP checks for `.lcpl` in `.epub` (which is embedded into the book during fulfillment process).

Reply:
that makes sense. could you add that description to the code? it’s makes things much clearer.